packages/ruleset/src/utils/validate-subschemas.js

```diff
@@ -0,0 +1,36 @@
+const validateSubschemas = (schema, path, validate) => {
+  const errors = [];
+  // invoke validation function
+  errors.push(...validate(schema, path));
+
+  // recursively process subschemas
```

Comment:
Are Spectral's resolved documents somehow protected from loops or do we need to do some loop protection of our own?

Reply:
Actually, I have no idea if Spectral protects its resolved documents from loops (I hope that they would).

However, I did add loop protection in the early days of this tool and we have test cases with intentionally placed circular references. Our existing loop protection doesn't apply to Spectral but I believe we run the Spectral rules on these test cases anyways. I think that's evidence that Spectral is doing some kind of protection, or those test cases should infinitely loop with these existing recursive validations. I can double check this.

Reply:
Okay I confirmed that there is protection, circular references are not an issue here 👍